cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h"
+
+#include "cartographer/common/make_unique.h"
+#include "cartographer/mapping/internal/optimization/cost_functions/cost_helpers.h"
+#include "gmock/gmock.h"
+#include "google/protobuf/text_format.h"
+
+namespace cartographer {
+namespace pose_graph {
+namespace {
+
+constexpr int kPoseDimension = 3;
+constexpr int kResidualsCount = 3;
+constexpr int kParameterBlocksCount = 2;
+constexpr int kJacobianColDimension = kResidualsCount * kPoseDimension;
+
+using ::google::protobuf::TextFormat;
+using ::testing::ElementsAre;
+using ResidualType = std::array<double, kResidualsCount>;
+using JacobianType = std::array<std::array<double, kJacobianColDimension>,
+                                kParameterBlocksCount>;
+
+class AutoDiffRelativePoseCost {
+ public:
+  explicit AutoDiffRelativePoseCost(
+      const proto::RelativePose2D::Parameters& parameters)
+      : translation_weight_(parameters.translation_weight()),
+        rotation_weight_(parameters.rotation_weight()),
+        first_T_second_(transform::ToRigid2(parameters.first_t_second())) {}
+
+  template <typename T>
+  bool operator()(const T* const start_pose, const T* const end_pose,
+                  T* e) const {
+    const std::array<T, 3> error = mapping::optimization::ScaleError(
+        mapping::optimization::ComputeUnscaledError(first_T_second_, start_pose,
```

Comment:
To me that's completely unclear why this is the same residual as the one in the implementation. Shouldn't we use this function as well to compute the residual in the first place?
Alternatively, could we use the residual only computation here? Maybe we can even extract a common pattern (I suppose that all cost functions could be tested this way). Let's maybe discuss this point.

Reply:
done